sklearn/utils/validation.py

```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
 
     if max_val is not None and x > max_val:
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
+
+
+def _check_sample_weight(sample_weight, y=None, n_samples=None, dtype=None,
+                         order=None, **kwargs):
+    """Validate sample weights
+
+    Parameters
+    ----------
+    sample_weight : {ndarray, Number or None}
+       input sample weights
+    y: ndarray or None
+       target variable. Either y or n_samples must be provided.
+    n_samples: int or None
+       expected number of samples. Either y or n_samples must be provided.
+    dtype: dtype
+       dtype of the validated sample_weight
+    order : 'F', 'C' or None (default=None)
+        Whether an array will be forced to be fortran or c-style.
+        When order is None (default), if ``sample_weights`` is an ndarray,
+        nothing is ensured about the memory layout of the output array,
+        otherwise it will be of 'C' order by default.
+
+    kwargs:
+       additional parameters to pass to check_array
+
+    Parameters
+    ----------
+    sample_weight : ndarray, shape=(n_samples,)
+       validated sample weights
+    """
+    if n_samples is not None and y is not None:
+        raise ValueError('Only one of y, n_samples must be provided!')
+    elif y is not None:
+        n_samples = y.shape[0]
+
+    if sample_weight is None or isinstance(sample_weight, numbers.Number):
+        if order is None:
```

Comment:
I don't see any reason for `sample_weight` to be a number. It's equivalent to all ones.
Only Ridge allows `sample_weight` to be a float, but I think it's useless and we should rather not allow a float as `sample_weight` and not support it here.

Reply:
I also don't see the point of it, but this PR aims to do a refactoring without breaking backward compatibility. If we are going to disallow it (and potentially break some user code) I would rather it be considered in a separate PR.